### PR TITLE
MINOR: Use explicit construction of clients in IntegrationTestHarness

### PR DESCRIPTION
**Pull request by hachikuji**
Pre-initialization of clients in IntegrationTestHarness is a cause of significant confusion and additionally has resulted in a bunch of inconsistent client creation patterns. This patch requires test cases to create needed clients explicitly and makes the creation logic more consistent.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/integration/kafka/api/IntegrationTestHarness.scala [hunk 1]
```diff
@@ -92,26 +84,67 @@ abstract class IntegrationTestHarness extends KafkaServerTestHarness {
...
+  def createProducer[K, V](acks: Int = -1,
+                           maxBlockMs: Long = 60 * 1000L,
+                           bufferSize: Long = 1024L * 1024L,
+                           retries: Int = Int.MaxValue,
+                           deliveryTimeoutMs: Int = 30 * 1000,
+                           lingerMs: Int = 0,
+                           requestTimeoutMs: Int = 20 * 1000,
+                           keySerializer: Serializer[K] = new ByteArraySerializer,
+                           valueSerializer: Serializer[V] = new ByteArraySerializer,
+                           configOverrides: Properties = new Properties()): KafkaProducer[K, V] = {
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/IntegrationTestHarness.scala]: This is kind of annoying. We should probably decouple getting the configs from creating the producer/consumer. Then we can avoid this issue, I think?

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala [hunk 2]
```diff
@@ -36,8 +36,6 @@ import org.apache.kafka.common.internals.Topic
 abstract class BaseConsumerTest extends IntegrationTestHarness {
```
Comment (omkreddy) [inline core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:35]: nit: unused import

File: core/src/test/scala/integration/kafka/api/CustomQuotaCallbackTest.scala [hunk 3]
```diff
@@ -48,8 +48,6 @@ class CustomQuotaCallbackTest extends IntegrationTestHarness with SaslSetup {
   override protected def interBrokerListenerName: ListenerName = new ListenerName("BROKER")
```
Comment (omkreddy) [inline core/src/test/scala/integration/kafka/api/CustomQuotaCallbackTest.scala:49]: nit: unused import

File: core/src/test/scala/unit/kafka/server/LogDirFailureTest.scala [hunk 4]
```diff
@@ -55,16 +55,6 @@ class LogDirFailureTest extends IntegrationTestHarness {
     createTopic(topic, partitionNum, serverCount)
```
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/server/LogDirFailureTest.scala:55]: nit: unused import

File: core/src/test/scala/integration/kafka/api/SaslMultiMechanismConsumerTest.scala [hunk 5]
```diff
@@ -46,20 +46,12 @@ class SaslMultiMechanismConsumerTest extends BaseConsumerTest with SaslSetup {
   @Test
```
Comment (omkreddy) [inline core/src/test/scala/integration/kafka/api/SaslMultiMechanismConsumerTest.scala:46]: nit: unused imports

File: core/src/test/scala/integration/kafka/api/SaslClientsWithInvalidCredentialsTest.scala [hunk 6]
```diff
@@ -74,11 +74,12 @@ class SaslClientsWithInvalidCredentialsTest extends IntegrationTestHarness with
```
Comment (omkreddy) [inline core/src/test/scala/integration/kafka/api/SaslClientsWithInvalidCredentialsTest.scala:73]: nit: unused import

File: core/src/test/scala/integration/kafka/api/IntegrationTestHarness.scala (hunk 1 above)
Comment (hachikuji) [reply core/src/test/scala/integration/kafka/api/IntegrationTestHarness.scala]: Ack. I think we should just avoid the optional arguments. They are rarely used anyway and they make this api confusing.

**Comment by rajinisivaram** [comment]
@hachikuji Looks like the test failures may be related to the changes in the PR.

**Comment by hachikuji** [comment]
@rajinisivaram Thanks, pushed an update which I think fixes the problem. One of the the nice things about this refactor is forcing some of these config dependencies to the surface.

File: core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala [hunk 7]
```diff
@@ -46,7 +44,7 @@ abstract class BaseQuotaTest extends IntegrationTestHarness {
   this.serverConfig.setProperty(KafkaConfig.GroupMinSessionTimeoutMsProp, "100")
   this.serverConfig.setProperty(KafkaConfig.GroupMaxSessionTimeoutMsProp, "30000")
   this.serverConfig.setProperty(KafkaConfig.GroupInitialRebalanceDelayMsProp, "0")
-  this.producerConfig.setProperty(ProducerConfig.ACKS_CONFIG, "0")
+  this.producerConfig.setProperty(ProducerConfig.ACKS_CONFIG, "-1")
```
Comment (rajinisivaram) [inline core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala:47]: Why was this change needed?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 8]
```diff
@@ -944,23 +969,28 @@ class PlaintextConsumerTest extends BaseConsumerTest {
...
+    sendRecords(producer, numRecords = 100, tp)
+    sendRecords(producer, numRecords = 100, tp2)
     val topic1 = "topic1"
-    val subscriptions = Set(tp, tp2) ++ createTopicAndSendRecords(topic1, 5, 100)
+    val subscriptions = Set(tp, tp2) ++ createTopicAndSendRecords(producer, topic1, 5, 100)
 
     // subscribe all consumers to all topics and validate the assignment
-    val consumerPollers = subscribeConsumers(consumers, List(topic, topic1))
 
+    val consumerGroup = Buffer[KafkaConsumer[Array[Byte], Array[Byte]]]()
```
Comment (rajinisivaram) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Since `consumerGroup` refers to the logical group, perhaps `consumersInGroup` or something along those lines would be less confusing?

File: core/src/test/scala/integration/kafka/api/TransactionsBounceTest.scala [hunk 9]
```diff
@@ -150,18 +147,13 @@ class TransactionsBounceTest extends KafkaServerTestHarness {
...
-
-    val consumer = TestUtils.createConsumer(TestUtils.getBrokerListStrFromServers(servers), groupId = groupId,
-      securityProtocol = SecurityProtocol.PLAINTEXT, props = Some(props))
+  private def createConsumerAndSubscribeToTopics(groupId: String,
+                                                 topics: List[String],
+                                                 readCommitted: Boolean = false) = {
+    val consumer = TestUtils.createConsumer(TestUtils.getBrokerListStrFromServers(servers),
+      groupId = groupId,
+      readCommitted = readCommitted,
+      enableAutoCommit = false)
```
Comment (rajinisivaram) [inline core/src/test/scala/integration/kafka/api/TransactionsBounceTest.scala:156]: `maxPollRecords` from above not needed?

File: core/src/test/scala/unit/kafka/server/DescribeLogDirsRequestTest.scala [hunk 10]
```diff
@@ -26,9 +26,9 @@ import org.junit.Test
 import java.io.File
 
 class DescribeLogDirsRequestTest extends BaseRequestTest {
+  logDirCount = 2
```
Comment (rajinisivaram) [inline core/src/test/scala/unit/kafka/server/DescribeLogDirsRequestTest.scala]: is there a reason why this is a `var` rather than a `val`?

**Review by rajinisivaram — APPROVED** [review]
@hachikuji Thanks for the PR, looks good. Left a few minor comments. I think there were build failures in the last  PR build because a new test was added that needs updating. Feel free to merge after fixing that since it will be hard to maintain the PR otherwise.

File: core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala (hunk 7 above)
Comment (hachikuji) [reply core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala:47]: The acks value provided here was being overridden by the default value in TestUtils.createProducer. Fixing this problem caused tests to fail and I couldn't think of a strong reason to use acks=0, so I just made the dependence explicit.

File: core/src/test/scala/integration/kafka/api/TransactionsBounceTest.scala (hunk 9 above)
Comment (hachikuji) [reply core/src/test/scala/integration/kafka/api/TransactionsBounceTest.scala:156]: I couldn't think of a reason why we'd need to override this. Typically the only case we do would do so is in order to decrease the default of 500.
